listenbrainz_spark/RabbitMQ/rabbitmq_connection.py

```diff
@@ -0,0 +1,61 @@
+import pika
+import logging
+from listenbrainz_spark import config
+import time
+import json
+import sys
+
+class SparkReader():
+    def __init__(self):
+        self.unique_ch = None
+        self.ERROR_RETRY_DELAY = 3
+
+    def connect_to_rabbitmq(self):
+        """Creates a RabbitMQ connection object
+        """
+        username = config.RABBITMQ['RABBITMQ_USERNAME']
+        password = config.RABBITMQ['RABBITMQ_PASSWORD']
+        host = config.RABBITMQ['RABBITMQ_HOST']
+        port = config.RABBITMQ['RABBITMQ_PORT']
+        virtual_host = config.RABBITMQ['RABBITMQ_VHOST']
+
+        while True:
+            try:
+                credentials = pika.PlainCredentials(username, password)
+                connection_parameters = pika.ConnectionParameters(
+                    host=host,
+                    port=port,
+                    virtual_host=virtual_host,
+                    credentials=credentials,
+                )
+                self.connection =  pika.BlockingConnection(connection_parameters)
+                break
+            except Exception as err:
+                error_message = "Cannot connect to RabbitMQ: {error}, retrying in {delay} seconds."
+                print(error_message.format(error=str(err), delay=self.ERROR_RETRY_DELAY))
+                time.sleep(self.ERROR_RETRY_DELAY)
+
+    def start(self, data):
+        """Publish data to RabbitMQ
+        """
+        if "RABBITMQ_HOST" not in config.RABBITMQ:
+            logging.critical("RabbitMQ service not defined. Sleeping {0} seconds and exiting.".format(self.ERROR_RETRY_DELAY))
+            time.sleep(self.ERROR_RETRY_DELAY)
+            sys.exit(-1)
+
+        try:
+            self.connect_to_rabbitmq()
```

Comment:
The next few lines get called for each time we write data to RabbitMQ. Each time we re-setup the connection to RabbitMQ, which takes time. I think it makes more sense to check to see if we have already established a connection (maybe test for unique_ch being not None?) and then try and use the connection right away. If the write to RabbitMQ fails, then re-do the connection.

So, I think this would best be split into two function:
- start() -- setup the connections, possibly re-setting up any timedout connections.
- write() -- write data to the connection, but if the write fails, call start() again?

Reply:
Yes. Sounds good:)